winit-core/src/window.rs

```diff
@@ -1530,6 +1630,185 @@ impl Default for ImePurpose {
     }
 }
 
+/// Request to send to IME.
+#[derive(Debug, PartialEq, Clone)]
+pub enum ImeRequest {
+    /// Enable the IME with the [`ImeCapabilities`] and [`ImeRequestData`] as initial state. When
+    /// the [`ImeRequestData`] is **not** matching capabilities fully, the default values will be
+    /// used instead.
+    ///
+    /// **Requesting to update data matching not enabled capabilities will result in update
+    /// being ignored.** The winit backend in such cases is recommended to log a warning. This
+    /// appiles to both [`ImeRequest::Enable`] and [`ImeRequest::Update`]. For details on
+    /// capabilities refer to [`ImeCapabilities`].
+    ///
+    /// To update the [`ImeCapabilities`], the IME must be disabled and then re-enabled.
+    Enable(ImeEnableRequest),
+    /// Update the state of already enabled IME. Issuing this request before [`ImeRequest::Enable`]
+    /// will result in error.
+    Update(ImeRequestData),
+    /// Disable the IME.
+    ///
+    /// **The disable request can not fail**.
+    Disable,
+}
+
+/// Initial IME request.
+#[derive(Debug, Clone, PartialEq)]
+pub struct ImeEnableRequest {
+    capabilities: ImeCapabilities,
+    request_data: ImeRequestData,
+}
+
+impl ImeEnableRequest {
+    /// Create request for the [`ImeRequest::Enable`]
+    ///
+    /// This will return [`None`] if some capability was requested but its initial value was not
+    /// set by the user.
+    pub const fn new(capabilities: ImeCapabilities, request_data: ImeRequestData) -> Option<Self> {
+        if capabilities.contains(ImeCapabilities::CURSOR_AREA) && request_data.cursor_area.is_none()
+        {
+            return None;
+        }
+
+        if capabilities.contains(ImeCapabilities::PURPOSE) && request_data.cursor_area.is_none() {
```

Comment:
`request_data.cursor_area.is_none()`

should be "purpose".

I think it would be easier to drop the `capabilities` argument because here you're checking the same thing twice.

Reply:
>Do you mean that you want the winit to advertise future properties that the user didn't opt into? The user won't be able to provide values for them anyway, so I must be misunderstading.

I meant here that future `ImeRequest::Update` won't work, if user won't pass `cursor_area`. But to figure that out and make backends more or less consistent, we need capabilities passed by the user rather than derived based off what they set initially, since in practice docs are not being read that often and you'll fix on a user report basis, while here we clearly communicate. 

I do understand that they are redundant though, it's just with something that generally lacks testing it's better to be more verbose.

Reply:
Got it, the redundancy is the point: it makes the user confirm the intention multiple times.